gtsam_unstable/geometry/Similarity3.cpp

```diff
@@ -85,10 +86,87 @@ Point3 Similarity3::transformFrom(const Point3& p, //
   return s_ * q;
 }
 
+Pose3 Similarity3::transformFrom(const Pose3& T) const {
+  Rot3 R = R_.compose(T.rotation());
+  Point3 t = Point3(s_ * (R_ * T.translation() + t_));
+  return Pose3(R, t);
+}
+
 Point3 Similarity3::operator*(const Point3& p) const {
   return transformFrom(p);
 }
 
+// Refer to: http://www5.informatik.uni-erlangen.de/Forschung/Publikationen/2005/Zinsser05-PSR.pdf Chapter 3
+Similarity3 Similarity3::AlignGivenR(const std::vector<Point3Pair>& abPointPairs, const Rot3& aRb) {
+  // calculate centroids
+  const Point3Pair centroids = mean(abPointPairs);
+  const Point3 aCentroid = centroids.first;
+  const Point3 bCentroid = centroids.second;
+
+  // calculate scale
+  double x = 0;
+  double y = 0;
+  Point3 aPoint, bPoint;
+  for (const Point3Pair& abPair : abPointPairs) {
+    std::tie(aPoint, bPoint) = abPair;
+    const Point3 da = aPoint - aCentroid;
+    const Point3 db = bPoint - bCentroid;
+    Vector3 Rdb = aRb * db;
+    y += da.transpose() * Rdb;
+    x += Rdb.transpose() * Rdb;
+  }
+  const double s = y / x;
+
+  // calculate translation
+  const Point3 aTb = (aCentroid - s * (aRb * bCentroid)) / s;
+  return Similarity3(aRb, aTb, s);
+}
+
+// Refer to: http://www5.informatik.uni-erlangen.de/Forschung/Publikationen/2005/Zinsser05-PSR.pdf Chapter 3
+Similarity3 Similarity3::Align(const std::vector<Point3Pair>& abPointPairs) {
+  const size_t n = abPointPairs.size();
+  if (n < 3) throw std::runtime_error("input should have at least 3 pairs of points");  // we need at least three pairs
+
+  // calculate centroids
+  const Point3Pair centroids = mean(abPointPairs);
+  const Point3 aCentroid = centroids.first;
+  const Point3 bCentroid = centroids.second;
+
+  // Add to form H matrix
+  Matrix3 H = Z_3x3;
+  Point3 aPoint, bPoint;
+  for (const Point3Pair& abPair : abPointPairs) {
+    std::tie(aPoint, bPoint) = abPair;
+    Point3 da = aPoint - aCentroid;
+    Point3 db = bPoint - bCentroid;
+    H += da * db.transpose();
+  }
+
+  // ClosestTo finds rotation matrix closest to H in Frobenius sense
+  Rot3 aRb = Rot3::ClosestTo(H);
+
+  return AlignGivenR(abPointPairs, aRb);
```

Comment:
*still* seems like a lot of code duplication, and, what's worse, double computation (centroids and H are basically computed twice). Think about a million points.